clustercontroller-core/src/main/java/com/yahoo/vespa/clustercontroller/core/ResourceExhaustionCalculator.java

```diff
@@ -0,0 +1,79 @@
+// Copyright Verizon Media. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
+package com.yahoo.vespa.clustercontroller.core;
+
+import com.yahoo.vdslib.state.Node;
+import com.yahoo.vespa.clustercontroller.core.hostinfo.HostInfo;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+/**
+ * Given a mapping of (opaque) resource names and their exclusive limits,
+ * this class acts as an utility to easily enumerate all the resources that
+ * a given node (or set of nodes) have exhausted.
+ */
+public class ResourceExhaustionCalculator {
+
+    private final boolean feedBlockEnabled;
+    private final Map<String, Double> feedBlockLimits;
+
+    public ResourceExhaustionCalculator(boolean feedBlockEnabled, Map<String, Double> feedBlockLimits) {
+        this.feedBlockEnabled = feedBlockEnabled;
+        this.feedBlockLimits = feedBlockLimits;
+    }
+
+    public ClusterStateBundle.FeedBlock inferContentClusterFeedBlockOrNull(Collection<NodeInfo> nodeInfos) {
+        if (!feedBlockEnabled) {
+            return null;
+        }
+        var exhaustions = enumerateNodeResourceExhaustionsAcrossAllNodes(nodeInfos);
+        if (exhaustions.isEmpty()) {
+            return null;
+        }
+        int maxDescriptions = 3;
+        String description = exhaustions.stream()
+                .limit(maxDescriptions)
+                .map(n -> String.format("%s on node %s (%.3g > %.3g)",
+                        n.resourceType, n.node.getIndex(),
+                        n.resourceUsage.getUsage(), n.limit))
```

Comment:
Resource types “attribute-enum-store” and “attribute-multi-value” will also contain a name. This should be part of the description message, and I guess added to ResourceUsage.

Reply:
Good point, I will add this as part of the follow-up work